farmerbot/models/config.go

```diff
@@ -0,0 +1,231 @@
+package models
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	"github.com/threefoldtech/tfgrid-sdk-go/farmerbot/constants"
+)
+
+// InputConfig is the inputs for configuration for farmerbot
+type InputConfig struct {
+	FarmID        uint32   `json:"farm_id" yaml:"farm_id" toml:"farm_id"`
+	IncludedNodes []uint32 `json:"included_nodes" yaml:"included_nodes" toml:"included_nodes"`
+	ExcludedNodes []uint32 `json:"excluded_nodes" yaml:"excluded_nodes" toml:"excluded_nodes"`
+	Power         Power    `json:"power" yaml:"power" toml:"power"`
+}
+
+// Config is the configuration data for farmerbot
+type Config struct {
+	Farm  substrate.Farm `json:"farm" yaml:"farm" toml:"farm"`
+	Nodes []Node         `json:"nodes" yaml:"nodes" toml:"nodes"`
+	Power Power          `json:"power" yaml:"power" toml:"power"`
+	m     sync.Mutex
+}
+
+// Set sets the config data from configuration inputs
+func (c *Config) Set(sub Sub, inputs InputConfig) error {
+	log.Debug().Msg("Set power")
+
+	if !reflect.DeepEqual(c.Power, inputs.Power) {
+		c.Power = inputs.Power
+	} else {
+		log.Debug().Msg("Configuration power didn't change")
+	}
+
+	// required from power for nodes
+	if c.Power.OverProvisionCPU == 0 {
+		c.Power.OverProvisionCPU = constants.DefaultCPUProvision
+	}
+
+	if c.Power.OverProvisionCPU < 1 || c.Power.OverProvisionCPU > 4 {
+		return fmt.Errorf("cpu over provision should be a value between 1 and 4 not %v", c.Power.OverProvisionCPU)
+	}
+
+	// set farm
+	log.Debug().Uint32("ID", inputs.FarmID).Msg("Set farm")
+	if inputs.FarmID != uint32(c.Farm.ID) {
+		farm, err := sub.GetFarm(inputs.FarmID)
+		if err != nil {
+			return err
+		}
+
+		c.Farm = *farm
+	} else {
+		log.Debug().Msg("Configuration farm didn't change")
+	}
+
+	// set nodes
+	err := c.SetConfigNodes(sub, inputs)
+	if err != nil {
+		return err
+	}
+
+	return c.validate()
+}
+
+// SetConfigNodes sets the config nodes from configuration inputs
+func (c *Config) SetConfigNodes(sub Sub, i InputConfig) error {
```

Comment:
Again while this is the wrong place, I also noticed that this only is done when the bot is starting. Which is wrong, instead the list of the nodes need (ids) need to be synced with the chain with every bot iteration.

Reply:
In other words, we need to make sure that the node ids the bot knows about must be exactly the same as the nodes (matching including/excluding lists) on the grid.